### ISSUE TITLE
Is there a reason the 16S databases doesn't contain species level information?

### ISSUE THREAD

**Issue by BioinformaticsBro** [DerrickWood/kraken2#164]

**Comment by jenniferlu717** [comment]
Its not a result of what Kraken 2 is doing. Its just that the 16S Silva and RDP databases dont specify species level although Greengenes does oddly.

**Comment by Marc-Ruebsam** [comment]
**Please reopen if you agree**

Correct me if I'm wrong but SILVA IS providing species level information. It is just that this information is discarded during the construction of the kraken2 database. Let me explain:

## Files ##
- [SILVA_132_SSURef_Nr99_tax_silva.fasta](https://ftp.arb-silva.de/release_132/Exports/SILVA_132_SSURef_Nr99_tax_silva.fasta.gz) (called _fasta file_ from now on)
- [tax_slv_ssu_132.txt](https://ftp.arb-silva.de/release_132/Exports/taxonomy/tax_slv_ssu_132.txt) (called _tax list_ from now on)
- [tax_slv_ssu_132.acc_taxid](https://ftp.arb-silva.de/release_132/Exports/taxonomy/tax_slv_ssu_132.acc_taxid) (called _seqid2taxid map_ from now on)

## An example: Lactobacillus fermentum ##
The fasta file contains 96 sequences of **_Lactobacillus fermentum_**.
```
$ grep "Lactobacillus fermentum" SILVA_132_SSURef_Nr99_tax_silva.fasta | cut -d" " -f2- | sort | uniq -c
     79 Bacteria;Firmicutes;Bacilli;Lactobacillales;Lactobacillaceae;Lactobacillus;Lactobacillus fermentum
      1 Bacteria;Firmicutes;Bacilli;Lactobacillales;Lactobacillaceae;Lactobacillus;Lactobacillus fermentum ATCC 14931
      5 Bacteria;Firmicutes;Bacilli;Lactobacillales;Lactobacillaceae;Lactobacillus;Lactobacillus fermentum CECT 5716
      5 Bacteria;Firmicutes;Bacilli;Lactobacillales;Lactobacillaceae;Lactobacillus;Lactobacillus fermentum F-6
      5 Bacteria;Firmicutes;Bacilli;Lactobacillales;Lactobacillaceae;Lactobacillus;Lactobacillus fermentum IFO 3956
      1 Bacteria;Firmicutes;Bacilli;Lactobacillales;Lactobacillaceae;Lactobacillus;Lactobacillus fermentum L930BB
```
All of which are assigned to **taxon ID 1837** in the seqid2taxid map.
```
$ grep "$(grep "Lactobacillus fermentum" SILVA_132_SSURef_Nr99_tax_silva.fasta | cut -d" " -f1 | sed 's#>##g')" tax_slv_ssu_132.acc_taxid | cut -f2 | sort | uniq -c
     96 1837
```
This taxon ID corresponds to genus **_Lactobacillus_** in the tax list.
```
$ grep -w 1837 tax_slv_ssu_132.txt
Bacteria;Firmicutes;Bacilli;Lactobacillales;Lactobacillaceae;Lactobacillus;     1837    genus           119
```
Hence, if we classify the sequences from the fasta file using its own database,
```
## extract L. fermentum sequences from fasta
$ samtools faidx SILVA_132_SSURef_Nr99_tax_silva.fasta $(grep "Lactobacillus fermentum" SILVA_132_SSURef_Nr99_tax_silva.fasta | cut -d" " -f1 | sed 's/>//g') | sed '/^[^>]/ y/U/T/' > Lactobacillus_fermentum_dummy_input.fasta

## build kraken2 database
$ kraken2-build --special silva --db silva_132

## classification with kraken2
$ kraken2 --use-names --db silva_132/ --output output --report report Lactobacillus_fermentum_dummy_input.fasta
Loading database information... done.
96 sequences (0.14 Mbp) processed in 0.024s (241.8 Kseq/m, 363.51 Mbp/m).
  96 sequences classified (100.00%)
  0 sequences unclassified (0.00%)
```
we see that all of them get classified as genus **_Lactobacillus_** with the **taxon ID 1837**, but not species **_Lactobacillus fermentum_**.
```
$ cat report
100.00  96      0       R       1       root
100.00  96      0       D       3         Bacteria
100.00  96      0       P       1672        Firmicutes
100.00  96      0       C       1673          Bacilli
100.00  96      0       O       1800            Lactobacillales
100.00  96      0       F       1836              Lactobacillaceae
100.00  96      96      G       1837                Lactobacillus
```
even though the species is clearly represented in the reference

## Tools ##
```
$ kraken2 --version
Kraken version 2.0.8-beta
Copyright 2013-2019, Derrick Wood (dwood@cs.jhu.edu)
$ samtools --version
samtools 1.9
Using htslib 1.9
Copyright (C) 2018 Genome Research Ltd.
```

**Comment by nico-chung** [comment]
There seems to be a similar issue the with the way RefSeq plasmids are labelled causing the plasmid name to be cut off during database construction. This makes it impossible for Kraken to identify the plasmids since it only gives the species the plasmid is from. 

Kraken version 2.0.8-beta

**Comment by jenniferlu717** [comment]
@Marc-Ruebsam Regarding SILVA, while yes, individual sequences may list a species in the FASTA file, Kraken2 uses both the SILVA sequences AND taxonomy structure. SILVA's taxonomy (the tax list file) does not contain a species-level, only genus level. 

@nico-chung please make a new issue, plasmids are different from the 16S databases and I'm not 100% sure what you are referring to.

**Comment by Marc-Ruebsam** [comment]
@jenniferlu717 Thank you for your reply. I am aware of the missing species assignment in SILVA, but struggle to understand the reasoning behind it. I contacted the SILVA-arb support regarding this question today.
In parallel I try to include the information from the [taxmap_embl_ssu_ref_nr99_132.txt](https://ftp.arb-silva.de/release_132/Exports/taxonomy/taxmap_embl_ssu_ref_nr99_132.txt.gz) to create a custom database with species associations. Just in case anyone is looking into this as well.

**Comment by Marc-Ruebsam** [comment]
Update:
I've build the database as mentioned above and now get species level classification for most of the sequences:
```
100.00  96      0       R       1       root
100.00  96      0       D       3         Bacteria
100.00  96      0       P       1672        Firmicutes
100.00  96      0       C       1673          Bacilli
100.00  96      0       O       1800            Lactobacillales
100.00  96      0       F       1836              Lactobacillaceae
100.00  96      27      G       1837                Lactobacillus
 65.62  63      63      S       16131837                      Lactobacillus fermentum
  2.08  2       2       S       2834414541                    Lactobacillus fermentum CECT 5716
  2.08  2       2       S       3343901837                    Lactobacillus fermentum IFO 3956
  2.08  2       2       S       3379564541                    Lactobacillus fermentum F-6
```

However, the implementation has a couple of compromises (e.g. strains are classifies as separate species). Hence, I wont make it a pull request. Happy to get feedback if anyone is interested.

**Comment by hempelc** [comment]
Hey Marc, it's been a while ago but I'm trying to do the same thing as you, building the database to get species-level identification. I was wondering what approach you chose to do that and if you're willing to share that with me?

**Comment by Marc-Ruebsam** [comment]
> Hey @Marc-Ruebsam, it's been a while ago but I'm trying to do the same thing as you, building the database to get species-level identification. I was wondering what approach you chose to do that and if you're willing to share it with me?

Hi @hempelc, some time has passed indeed. As I remember it, I ended up implementing both, a genus level annotation as well as a "best possible" custom sepecies level annotation. You can find the [Snakemake pipeline here on Github](https://github.com/Marc-Ruebsam/MeBaPiNa). I've attached a copy of my master thesis describing the project. The methods section could be helpful.

[Marc_Ruebsam_LayoutKorrektur_Online.pdf](https://github.com/DerrickWood/kraken2/files/6828490/Marc_Ruebsam_LayoutKorrektur_Online.pdf)

**Comment by hempelc** [comment]
Amazing, thanks for the very thorough reply! I couldn't have asked for better help.
Have a great day!

**Comment by shreyanobelex** [comment]
@Marc-Ruebsam I have trying to rebuild the database using the .txt file you have provided here. But it is still showing the same output without species ! what script did you use to build it ? Can you please provide it?

**Comment by Marc-Ruebsam** [comment]
> @Marc-Ruebsam I have been trying to rebuild the database using the .txt file you have provided here. But it is still showing the same output without species ! what script did you use to build it ? Can you please provide it?

Hi @shreyanobelex , did you have a look at the repo I mentioned in my response above? It containes all the scripts I have used.
